In [5]:
!wget http://dinamica10.fciencias.unam.mx/tmp/t2_1C_PB1-11_e41-81_mod_noise-04_451000.pat.txt

--2020-05-02 20:32:25--  http://dinamica10.fciencias.unam.mx/tmp/t2_1C_PB1-11_e41-81_mod_noise-04_451000.pat.txt
Resolving dinamica10.fciencias.unam.mx (dinamica10.fciencias.unam.mx)... 132.248.181.116
Connecting to dinamica10.fciencias.unam.mx (dinamica10.fciencias.unam.mx)|132.248.181.116|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 235012634 (224M) [text/plain]
Saving to: ‘t2_1C_PB1-11_e41-81_mod_noise-04_451000.pat.txt’

t2_1C_PB1-11_e41-81 100%[===================>] 224.12M  11.1MB/s    in 21s     

2020-05-02 20:32:46 (10.9 MB/s) - ‘t2_1C_PB1-11_e41-81_mod_noise-04_451000.pat.txt’ saved [235012634/235012634]



In [0]:
#Importamos librerias utiles
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, accuracy_score
import tensorflow as tf
from keras.models import Sequential
from keras.layers import Dense, Activation, Input, Dropout, Convolution2D, MaxPooling2D
from sklearn.preprocessing import LabelEncoder
from tensorflow import keras


capas = keras.layers

In [0]:
#leemos el archivo que acabamos de descargar, omitiendo las primeras tres filas ya que esas no corresponden a los datos
datos = pd.read_csv("t2_1C_PB1-11_e41-81_mod_noise-04_451000.pat.txt", delimiter=';', skiprows=3,header=None)


In [8]:
z = np.where(datos[0]==0)
z[0].shape
#La primera columna del dataframe contiene puros ceros por lo que la vamos a eliminar

(451000,)

In [9]:
#tiramos la primera columna ya que son puros ceros y la columna 82 ya que eso se nos pidio en la tarea
datos = datos.drop([0,82], axis=1)
datos.rename(columns={81: "Clase"},inplace = True)
datos.sample(6)

1        2        3        4  ...       78       79       80  Clase
417962  0.00481  0.01008  0.01466  0.02004  ...  0.00000  0.00000  0.00000    418
14165   0.00935  0.01978  0.02987  0.04093  ...  0.00000  0.00000  0.00000     15
425261  0.00514  0.00965  0.01483  0.02192  ...  0.00000  0.00000  0.00000    426
232150  0.00754  0.01627  0.02045  0.02945  ...  0.00000  0.00000  0.00000    233
369692  0.00557  0.01025  0.01553  0.02198  ...  0.18212  0.17284  0.16601    370
407373  0.00506  0.01168  0.01488  0.02212  ...  0.00000  0.00000  0.00000    408

[6 rows x 81 columns]

In [10]:
#contamos cuantas clases hay
print(datos.max(axis=0))
print(datos.min(axis=0))
#en total son 451 clases
datos['Clase'].value_counts()
#aparentemente las clases estan balanceadas.

1          0.01174
2          0.02326
3          0.03314
4          0.04665
5          0.08597
           ...    
77         0.35525
78         0.34022
79         0.34467
80         0.33663
Clase    451.00000
Length: 81, dtype: float64
1        0.00414
2        0.00810
3        0.01173
4        0.01628
5        0.03051
          ...   
77       0.00000
78       0.00000
79       0.00000
80       0.00000
Clase    1.00000
Length: 81, dtype: float64


255    1000
312    1000
327    1000
72     1000
328    1000
       ... 
408    1000
153    1000
409    1000
154    1000
256    1000
Name: Clase, Length: 451, dtype: int64

In [11]:
#dividimos datos en entrenamiento y prueba
entrenamiento, prueba = train_test_split(datos, test_size=0.25)
print(entrenamiento['Clase'].value_counts())
print(prueba['Clase'].value_counts())

280    785
124    784
282    782
302    781
258    780
      ... 
333    719
209    718
442    716
248    716
413    715
Name: Clase, Length: 451, dtype: int64
413    285
248    284
442    284
209    282
333    281
      ... 
258    220
302    219
282    218
124    216
280    215
Name: Clase, Length: 451, dtype: int64


In [0]:
#definimos conjuntos de entrenamiento y prueba (x e y)
xTrain = (entrenamiento.drop(columns=['Clase'])).to_numpy()
yTrain = (entrenamiento['Clase']).to_numpy()
xTest = (prueba.drop(columns=['Clase'])).to_numpy()
yTest = (prueba['Clase']).to_numpy()

In [53]:
#con una sola neurona de salida no es posible usar la funcion de perdida entropia cruzada categorica
#Este codigo se comenta porque arroja error, pero se deja para dejar el historico
#modeloS = Sequential()
#modeloS.add(Dense(80, input_shape=(80,),activation='relu'))
#modeloS.add(Dense(1000,activation='relu'))
#modeloS.add(Dense(1000,activation='relu'))
#modeloS.add(Dense(1,activation='softmax'))

#modeloS.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
#modeloS.fit(xTrain,yTrain,epochs=5,batch_size=256)

Epoch 1/5


InvalidArgumentError: ignored

In [106]:
#Para tener una sola neurona de salida, es usamos la funcion de perdida mse. Sin embargo el rendimiento de este modelo no es nada bueno
modelo = Sequential()
#80 neuronas capa de entrada con 80 atributos de entrada para cada una
modelo.add(Dense(80, input_shape=(80,),activation='linear'))
#1000 neuronas totalmente conectadas con las siguientes mil y las neuronas de entrada. estas mil neuronas tienen funcion de activacion relu
modelo.add(Dense(1000,activation='relu'))
#1000 neuronas totalmente conectadas con las siguiente  y las neuronas anteriores. estas mil neuronas tienen funcion de activacion relu
modelo.add(Dense(1000,activation='relu'))
#una neurona de salida con funcion de activacion lineal
modelo.add(Dense(1,activation='linear'))

#compila modelo
modelo.compile(optimizer='adam', loss='mse', metrics=['accuracy'])
#entrena red neuronal
modelo.fit(xTrain,yTrain,epochs=5,batch_size=128)

Epoch 1/5
338250/338250 [==============================] - 47s 140us/step - loss: 1940.5278 - accuracy: 0.0738
Epoch 2/5
338250/338250 [==============================] - 48s 141us/step - loss: 21.1597 - accuracy: 0.1138
Epoch 3/5
338250/338250 [==============================] - 48s 141us/step - loss: 18.8714 - accuracy: 0.1349
Epoch 4/5
338250/338250 [==============================] - 47s 140us/step - loss: 18.4499 - accuracy: 0.1403
Epoch 5/5
338250/338250 [==============================] - 47s 140us/step - loss: 16.5262 - accuracy: 0.1474


In [74]:
#ahora si ajustamos que en la capa de salida hayan 452 neuronas(clases) es posible entrenr usando la funcion de perdida
#entropia categorica y ademas el rendimiento del modelo es bastante bueno, habria que ver que sucede con los datos de prueba
modeloS = Sequential()
modeloS.add(Dense(80, input_shape=(80,),activation='linear'))
modeloS.add(Dense(2000,activation='relu'))
modeloS.add(Dense(2000,activation='relu'))
modeloS.add(Dense(452,activation='softmax'))

modeloS.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
modeloS.fit(xTrain,yTrain,epochs=5,batch_size=256)


Epoch 1/5
338250/338250 [==============================] - 138s 407us/step - loss: 1.0541 - accuracy: 0.7075
Epoch 2/5
338250/338250 [==============================] - 139s 412us/step - loss: 0.1161 - accuracy: 0.9673
Epoch 3/5
338250/338250 [==============================] - 138s 408us/step - loss: 0.0496 - accuracy: 0.9870
Epoch 4/5
338250/338250 [==============================] - 138s 407us/step - loss: 0.0355 - accuracy: 0.9905
Epoch 5/5
338250/338250 [==============================] - 138s 408us/step - loss: 0.0402 - accuracy: 0.9893


In [104]:
prediccion = modeloS.predict(xTest)
#define un nuevo vector a partir de las probabilidades tal que se redondee cada valor (0 o 1). esto se hace para poder calcular las metricas correctamente 
prediccion = np.rint(prediccion)
#Dado que el prediccion es una matriz y no un vector debemos transformar los valores predichos a sus respectivas clases esto se hace buscando la posicion de
#la clase mas probable (1) y se guarda un un vector llamado prediccion1 la posicion en la cual se encuentra dicha clase.
prediccion1 = np.zeros((prediccion.shape[0]),dtype=int)
for i in range(prediccion.shape[0]):
  z = np.where(prediccion[i]==1)
  prediccion1[i] = z[0][0]

print(accuracy_score(prediccion1,yTest))
#imprime tabla con los scores precision, recall, f1
print(classification_report(prediccion1,yTest))

0.999840354767184
              precision    recall  f1-score   support

           1       1.00      1.00      1.00       241
           2       1.00      1.00      1.00       259
           3       1.00      1.00      1.00       247
           4       1.00      1.00      1.00       247
           5       1.00      1.00      1.00       223
           6       1.00      1.00      1.00       246
           7       1.00      1.00      1.00       235
           8       1.00      1.00      1.00       266
           9       1.00      1.00      1.00       262
          10       1.00      1.00      1.00       253
          11       1.00      1.00      1.00       258
          12       1.00      1.00      1.00       257
          13       1.00      1.00      1.00       262
          14       1.00      1.00      1.00       257
          15       1.00      1.00      1.00       228
          16       1.00      1.00      1.00       248
          17       1.00      1.00      1.00       245
         

In [0]:
#hacemos one hot encoiding a los datos y de entrenamiento para 452 clases
one_hot_labels = keras.utils.to_categorical(yTrain, num_classes=452)
#hacemos one hot encoiding a los datos y de prueba para 452 clases
labelsYVA = keras.utils.to_categorical(yTest, num_classes=452)
#con estos vectores ahora se va a entrenar una red neuronal

In [56]:
#red neuronal de prueba con dos capas, la de entrada tiene 80 neuronas y salida tiene 452 neuronas que corresponden a la clase
model = Sequential()
model.add(Dense(80, activation='relu', input_dim=80))
model.add(Dense(452, activation='softmax'))
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train the model, iterating on the data in batches of 32 samples
model.fit(xTrain, one_hot_labels, epochs=10, batch_size=32)

Epoch 1/10
338250/338250 [==============================] - 18s 52us/step - loss: 2.7287 - accuracy: 0.2903
Epoch 2/10
338250/338250 [==============================] - 17s 50us/step - loss: 1.2043 - accuracy: 0.5651
Epoch 3/10
338250/338250 [==============================] - 17s 50us/step - loss: 0.9478 - accuracy: 0.6376
Epoch 4/10
338250/338250 [==============================] - 17s 50us/step - loss: 0.8212 - accuracy: 0.6791
Epoch 5/10
338250/338250 [==============================] - 17s 50us/step - loss: 0.7280 - accuracy: 0.7139
Epoch 6/10
338250/338250 [==============================] - 17s 51us/step - loss: 0.6511 - accuracy: 0.7444
Epoch 7/10
338250/338250 [==============================] - 17s 51us/step - loss: 0.5804 - accuracy: 0.7744
Epoch 8/10
338250/338250 [==============================] - 17s 51us/step - loss: 0.5167 - accuracy: 0.8023
Epoch 9/10
338250/338250 [==============================] - 17s 51us/step - loss: 0.4623 - accuracy: 0.8271
Epoch 10/10
338250/338250 [=

In [59]:
#red neuronal con dos capas ocultas de 2000 neuronas cada una. Una de entrada con 80 neuronas y una capa de salida con 452 neuronas que corresponden a cada clase
#Optimizador RMSProp
model = Sequential()
model.add(Dense(80, activation='relu', input_dim=80))
model.add(Dense(2000,activation='relu'))
model.add(Dense(2000,activation='relu'))
model.add(Dense(452, activation='softmax'))
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# entrena modelo con 5 epocas
model.fit(xTrain, one_hot_labels, epochs=5, batch_size=32)

Epoch 1/5
338250/338250 [==============================] - 435s 1ms/step - loss: 0.9805 - accuracy: 0.6402
Epoch 2/5
338250/338250 [==============================] - 426s 1ms/step - loss: 0.2790 - accuracy: 0.9347
Epoch 3/5
338250/338250 [==============================] - 422s 1ms/step - loss: 0.2206 - accuracy: 0.9589
Epoch 4/5
338250/338250 [==============================] - 429s 1ms/step - loss: 0.1894 - accuracy: 0.9661
Epoch 5/5
338250/338250 [==============================] - 418s 1ms/step - loss: 0.1749 - accuracy: 0.9696


In [0]:
#predice las probabilidades de cada x de prueba
predicted = model.predict(xTest)

In [65]:
#define un nuevo vector a partir de las probabilidades tal que se redondee cada valor (0 o 1). esto se hace para poder calcular las metricas correctamente 
g= np.rint(predicted)
#imprime el accuracy score
print(accuracy_score(g,labelsYVA))
#imprime tabla con los scores precision, recall, f1
print(classification_report(g,labelsYVA))

0.9995920177383592
              precision    recall  f1-score   support

           0       0.00      0.00      0.00         0
           1       1.00      1.00      1.00       242
           2       1.00      1.00      1.00       259
           3       1.00      1.00      1.00       248
           4       1.00      1.00      1.00       247
           5       1.00      0.94      0.97       238
           6       1.00      1.00      1.00       246
           7       1.00      1.00      1.00       235
           8       1.00      1.00      1.00       266
           9       1.00      1.00      1.00       263
          10       1.00      1.00      1.00       253
          11       1.00      1.00      1.00       259
          12       1.00      1.00      1.00       257
          13       1.00      0.98      0.99       266
          14       1.00      1.00      1.00       256
          15       1.00      1.00      1.00       228
          16       1.00      1.00      1.00       248
        

/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in samples with no true labels. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [109]:
#red neuronal con dos capas ocultas de 2000 neuronas cada una. Una de entrada con 80 neuronas y una capa de salida con 452 neuronas que corresponden a cada clase
#dado que cada epoca tarda aproximadamente 10 minutos en terminar, reduciremos a 3 epocas. Para este caso usamos el optimizador Adadelta
model = Sequential()
model.add(Dense(80, activation='relu', input_dim=80))
model.add(Dense(2000,activation='relu'))
model.add(Dense(2000,activation='relu'))
model.add(Dense(452, activation='softmax'))
model.compile(optimizer='Adadelta',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# entrena modelo con 3 epocas
model.fit(xTrain, one_hot_labels, epochs=3, batch_size=32)

Epoch 1/3
338250/338250 [==============================] - 541s 2ms/step - loss: 1.1074 - accuracy: 0.6612
Epoch 2/3
338250/338250 [==============================] - 537s 2ms/step - loss: 0.1322 - accuracy: 0.9658
Epoch 3/3
338250/338250 [==============================] - 539s 2ms/step - loss: 0.0747 - accuracy: 0.9826


In [110]:
#predice las probabilidades de cada x de prueba
predicted = model.predict(xTest)
#define un nuevo vector a partir de las probabilidades tal que se redondee cada valor (0 o 1). esto se hace para poder calcular las metricas correctamente 
g= np.rint(predicted)
#imprime el accuracy score
print(accuracy_score(g,labelsYVA))
#imprime tabla con los scores precision, recall, f1
print(classification_report(g,labelsYVA))

0.9998669623059867
              precision    recall  f1-score   support

           0       0.00      0.00      0.00         0
           1       1.00      1.00      1.00       242
           2       1.00      1.00      1.00       259
           3       1.00      1.00      1.00       248
           4       1.00      1.00      1.00       247
           5       1.00      1.00      1.00       223
           6       1.00      1.00      1.00       246
           7       1.00      1.00      1.00       235
           8       1.00      1.00      1.00       266
           9       1.00      1.00      1.00       262
          10       1.00      1.00      1.00       253
          11       1.00      1.00      1.00       259
          12       1.00      1.00      1.00       257
          13       1.00      1.00      1.00       262
          14       1.00      1.00      1.00       256
          15       1.00      1.00      1.00       228
          16       1.00      1.00      1.00       248
        

/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [111]:
#red neuronal con dos capas ocultas de 2000 neuronas cada una. Una de entrada con 80 neuronas y una capa de salida con 452 neuronas que corresponden a cada clase
#dado que cada epoca tarda aproximadamente 10 minutos en terminar, reduciremos a 3 epocas. Para este caso usamos el optimizador descenso del gradiente estocastico SDG
model = Sequential()
model.add(Dense(80, activation='relu', input_dim=80))
model.add(Dense(2000,activation='relu'))
model.add(Dense(2000,activation='relu'))
model.add(Dense(452, activation='softmax'))
model.compile(optimizer='SGD',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# entrena modelo con 3 epocas
model.fit(xTrain, one_hot_labels, epochs=3, batch_size=32)

Epoch 1/3
338250/338250 [==============================] - 284s 839us/step - loss: 5.3207 - accuracy: 0.0141
Epoch 2/3
338250/338250 [==============================] - 280s 829us/step - loss: 3.2665 - accuracy: 0.1327
Epoch 3/3
338250/338250 [==============================] - 279s 824us/step - loss: 1.8823 - accuracy: 0.3571


In [112]:
#predice las probabilidades de cada x de prueba
predicted = model.predict(xTest)
#define un nuevo vector a partir de las probabilidades tal que se redondee cada valor (0 o 1). esto se hace para poder calcular las metricas correctamente 
g= np.rint(predicted)
#imprime el accuracy score
print(accuracy_score(g,labelsYVA))
#imprime tabla con los scores precision, recall, f1
print(classification_report(g,labelsYVA))

0.20923281596452328
              precision    recall  f1-score   support

           0       0.00      0.00      0.00         0
           1       0.00      0.00      0.00         0
           2       0.00      0.00      0.00         0
           3       0.00      0.00      0.00         0
           4       0.00      0.00      0.00         0
           5       0.00      0.00      0.00         0
           6       0.49      1.00      0.66       121
           7       0.00      0.00      0.00         0
           8       0.00      0.00      0.00         0
           9       0.92      0.79      0.85       302
          10       0.00      0.00      0.00         0
          11       1.00      0.49      0.66       527
          12       0.00      0.00      0.00         0
          13       0.00      0.00      0.00         0
          14       0.00      0.00      0.00         0
          15       1.00      0.54      0.70       422
          16       0.00      0.00      0.00         0
       

/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in samples with no true labels. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


Observamos que al usar el optimizador SDG para tres epocas, se obtiene un accuracy de 0.35 y una perdida de 1.88. Para los datos de prueba para este caso se obtiene un acuracy del 0.21.

Al intercambiar el optimizador por Adadelta se obtiene un acuracy de 0.9826 y una perdida de 0.07. Para los datos de prueba se obtiene un acuracy de 0.99986.

Al emplear el optimizador RMSprop se obtiene un acuracy para tres epocas de 0.9589 con una perdida de 0.22. Para los datos de prueba un acuracy de 0.99959.
Con esto vemos que el optimizador que fue mejor fue el Adadelta.
Notemos que con el SDG tenemos una perdida relaticamente grade y quizá al corrrer más epocas se podria llegar a obtener un rendimiento mejor.
